packages/techdocs-common/src/stages/publish/local.ts

```diff
@@ -0,0 +1,144 @@
+import fs from 'fs-extra';
+import { Logger } from 'winston';
+import { Entity, EntityName } from '@backstage/catalog-model';
+import {
+  resolvePackagePath,
+  PluginEndpointDiscovery,
+} from '@backstage/backend-common';
+import { Config } from '@backstage/config';
+import { PublisherBase, PublishRequest } from './types';
+
+/* `remoteUrl` is the URL which serves files from the local publisher's static directory. */
+export type LocalPublishReturn = Promise<{ remoteUrl: string }>;
+
+const staticDocsDir = resolvePackagePath(
+  '@backstage/plugin-techdocs-backend',
+  'static/docs',
+);
+
+/**
+ * Local publisher which uses the local filesystem to store the generated static files. It uses a directory
+ * called "static" at the root of techdocs-backend plugin.
+ */
+export class LocalPublish implements PublisherBase {
+  private readonly config: Config;
+  private readonly logger: Logger;
+  private readonly discovery: PluginEndpointDiscovery;
+
+  constructor(
+    config: Config,
+    logger: Logger,
+    discovery: PluginEndpointDiscovery,
+  ) {
+    this.config = config;
+    this.logger = logger;
+    this.discovery = discovery;
+  }
+
+  publish({ entity, directory }: PublishRequest): LocalPublishReturn {
+    const entityNamespace = entity.metadata.namespace ?? 'default';
+
+    const publishDir = resolvePackagePath(
+      '@backstage/plugin-techdocs-backend',
+      'static/docs',
+      entityNamespace,
+      entity.kind,
+      entity.metadata.name,
+    );
+
+    if (!fs.existsSync(publishDir)) {
+      this.logger.info(`Could not find ${publishDir}, creating the directory.`);
+      fs.mkdirSync(publishDir, { recursive: true });
+    }
+
+    return new Promise((resolve, reject) => {
+      fs.copy(directory, publishDir, err => {
+        if (err) {
+          this.logger.debug(
+            `Failed to copy docs from ${directory} to ${publishDir}`,
+          );
+          reject(err);
+        }
+
+        this.discovery
+          .getBaseUrl('techdocs')
+          .then(techdocsApiUrl => {
+            resolve({
+              remoteUrl: `${techdocsApiUrl}/static/docs/${entity.metadata.name}`,
+            });
+          })
+          .catch(reason => {
+            reject(reason);
+          });
+      });
+    });
+  }
+
+  fetchTechDocsMetadata(entityName: EntityName): Promise<string> {
+    return new Promise((resolve, reject) => {
+      this.discovery.getBaseUrl('techdocs').then(techdocsApiUrl => {
+        const storageUrl = new URL(
+          new URL(this.config.getString('techdocs.storageUrl')).pathname,
+          techdocsApiUrl,
+        ).toString();
+
+        const entityRootDir = `${entityName.namespace}/${entityName.kind}/${entityName.name}`;
+        const metadataURL = `${storageUrl}/${entityRootDir}/techdocs_metadata.json`;
+        fetch(metadataURL)
+          .then(response =>
+            response
+              .json()
+              .then(techdocsMetadataJson => resolve(techdocsMetadataJson))
```

Comment:
See similar note on GCS publisher: seems like it'd be useful to resolve a typed object here instead of a string.

Reply:
> Yes, agreed! Currently there is no type definition for TechDocs Metadata and there should be one. I'll follow up by creating an issue about it.